### PR TITLE
major api, model, client and demoui changes. Supporting Hybrid index.

### PR DESCRIPTION
**Pull request by iRhysBradbury**
There's quite a bit going on here.

Refactoring nomenclature of model - I found it really hard to follow, so ive simplified it all.

Removing the notion of a nag and paf index, now only hybrid.

Removing unused code

Adding hybrid index query

API format param now optional

Changing the response model to cope with ES model

Refactoring pagination DSL

Refactoring sourceExcludes DSL

Refactoring ES Controller health check

Fixing Tests where applicable

Removing/ignoring redundant tests

We need to retest the server

### CODE REVIEW

**Comment by iRhysBradbury** [comment]
Codacy is complaining about nothing important.

**Comment by iRhysBradbury** [comment]
@Mironor That's correct.
 - Fixed the ones I could
- Removed the ones which are testing functionality we don't need (because that functionality has been removed)
-  Ignored some tests but kept the `should` line because I couldn't understand them

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  saoEndSuffix: String,
+  level: String,
+  officialFlag: String,
+  logicalStatus: String,
+  streetDescriptor: String,
+  townName: String,
+  locality: String
+) extends Formattable {
+
+  def formatAddress: String = {
```
Comment (Mironor) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala]: Would it be possible to return those methods where they were, outside of the DTO?

File: tagger/README.md [hunk 2]
```diff
@@ -35,21 +35,22 @@ $ xcode-select --install
 ```
 $ cd /path/to/address-index-api/tagger/crftagger
 
-# build liblbfgs first
```
Comment (iRhysBradbury) [inline tagger/README.md]: Merge with develop

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html [hunk 3]
```diff
@@ -34,6 +34,8 @@
 
         <link rel="stylesheet" media="screen" href='@routes.Assets.versioned("stylesheets/ons.css")'>
         <link rel="shortcut icon" type="image/png" href='@routes.Assets.versioned("images/favicon.ico")'>
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.0/jquery.min.js"></script>
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/main.scala.html]: SingleMatchUI ajax

File: server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala [hunk 4]
```diff
@@ -1,167 +1,117 @@
...
+import uk.gov.ons.addressIndex.model.{AddressScheme, BritishStandard7666, PostcodeAddressFile}
+import uk.gov.ons.addressIndex.model.db.index.HybridIndex
+import uk.gov.ons.addressIndex.server.modules.Model._
 import scala.concurrent.{ExecutionContext, Future}
 
+object Model {
+  case class Pagination(offset: Int, limit: Int)
+  implicit class AutoPaginate(searchDefinition: SearchDefinition) {
+    def paginate(implicit p: Pagination): SearchDefinition = {
+      searchDefinition start p.offset limit p.limit
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala]: This is pure infix notation :)

File: server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala [hunk 5]
```diff
@@ -1,167 +1,117 @@
...
+    }
+  }
+
+  implicit class AutoSource(searchDefinition: SearchDefinition) {
+    def format(implicit optFmt: Option[AddressScheme]) = {
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala]: It is implicit so the useage fits with the ElasticDSL.
(you dont have to add params in yourself, just apply the format which you have)
Checkout the impl lines:

`search.in(index).format`

keeps things clean

File: server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala [hunk 6]
```diff
@@ -1,167 +1,117 @@
...
+    * @param info
+    * @param t
+    * @param executable
+    * @tparam T
+    * @tparam R
+    * @tparam Q
+    * @return
     */
-  def queryNagAddresses(start: Int, limit: Int, tokens: Seq[CrfTokenResult]) : Future[NationalAddressGazetteerAddresses]
+  def logExecute[T, R, Q](info: String)(t: T)(implicit executable: Executable[T, R, Q]): Future[Q] = {
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala]: This is a wrapper on the Elastic4s Client, nothing fancy going on here just a copy of the Elastic4s's Client.execute method 

The idea is that you use this method instead, and we log the elastic4s `json` dynamically

also the DSL looks nice :)

File: server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala [hunk 7]
```diff
@@ -1,167 +1,117 @@
...
-import uk.gov.ons.addressIndex.parsers.Tokens
-import uk.gov.ons.addressIndex.parsers.Tokens.Token
-
+import uk.gov.ons.addressIndex.crfscala.CrfScala.CrfTokenResult
+import uk.gov.ons.addressIndex.model.{AddressScheme, BritishStandard7666, PostcodeAddressFile}
+import uk.gov.ons.addressIndex.model.db.index.HybridIndex
+import uk.gov.ons.addressIndex.server.modules.Model._
 import scala.concurrent.{ExecutionContext, Future}
 
+object Model {
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala]: Good spot, this needs to be moved

File: server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala [hunk 8]
```diff
@@ -1,167 +1,117 @@
...
-    * @param uprn the identificator of the address
-    * @return Fucture containing a NAG address or `None` if not in the index
+    * @param tokens
+    * @param p
+    * @param fmt
+    * @return
     */
-  def queryNagUprn(uprn : String) : Future[Option[NationalAddressGazetteerAddress]]
+  def queryAddress(tokens: Seq[CrfTokenResult])
+    (implicit p: Pagination, fmt: Option[AddressScheme]): Future[RichSearchResponse] = {
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/modules/AddressIndexRepository.scala]: The underlying usage required implicit params, I could move them to the underlying usage.
The idea is that you don't have to worry about the format and pagination, let the scala compiler do the magic :)

File: server/app/uk/gov/ons/addressIndex/server/controllers/AddressController.scala [hunk 9]
```diff
@@ -1,106 +1,131 @@
...
+    //get the defaults and maxima for the paging parameters from the config
     val defLimit = conf.config.elasticSearch.defaultLimit
     val defOffset = conf.config.elasticSearch.defaultOffset
     val maxLimit = conf.config.elasticSearch.maximumLimit
     val maxOffset = conf.config.elasticSearch.maximumOffset
-// TODO Look at refactoring to use types
-    val limval = limit.getOrElse(defLimit.toString())
-    val offval = offset.getOrElse(defOffset.toString())
+
+    //TODO REMOVE This
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/controllers/AddressController.scala]: Should remove the code below and use the scala type system.
specifically;
limit: String  -> limit: Int
offset: String -> offset: Int

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/PostcodeAddressFile.scala [hunk 10]
```diff
@@ -0,0 +1,89 @@
+package uk.gov.ons.addressIndex.model.db.index
+
+import uk.gov.ons.addressIndex.model.server.response.{PAF, PAFWithFormat}
+
+trait Formattable {
+  def delimitByComma(parts: String*) = parts.map(_.trim).filter(_.nonEmpty).mkString(", ")
+}
+
+case class PostcodeAddressFile(
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/PostcodeAddressFile.scala]: Should we wrap it with a context OBJ? 
```
object Address {
 case class PostcodeAddressFileAddress(blah: Type)
}
```

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    val pao = if (paoText == organisation) paoNumbers else s"$paoText, $paoNumbers"
+
+    val trimmedStreetDescriptor = streetDescriptor.trim
+    val buildingNumberWithStreetDescription = s"$pao $trimmedStreetDescriptor"
+
+    delimitByComma(organisation, sao, buildingNumberWithStreetDescription, locality,
+      townName, postcodeLocator)
+  }
+
+  def toNagWithFormat: NAGWithFormat = {
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala]: 23 fields :(

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala (hunk 1 above)
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala]: Checkout the formattable trait; moved the common repeated logic to one place.
Makes things easier to follow IMO, standard POJO-ee-ness.

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 12]
```diff
@@ -1,322 +1,196 @@
...
-  * @param response found content
-  * @param status   response status / message
-  * @param errors   encountered errors (or an empty list if there is no errors)
-  */
-case class AddressByUprnResponseContainer(
-  response: AddressByUprnResponse,
-  status: AddressResponseStatus,
-  errors: Seq[AddressResponseError] = Seq.empty[AddressResponseError]
-)
+object Model {
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: Good spot - needs to be moved

File: client/src/main/scala/uk/gov/ons/addressIndex/client/AddressIndexClient.scala [hunk 13]
```diff
@@ -61,23 +63,33 @@ trait AddressIndexClient {
...
 
   implicit class AddressIndexPathToWsAugmenter(p: AddressIndexPath)
     (implicit client: WSClient, host: AddressIndexServerHost) {
     def toReq(): WSRequest = {
-      client url s"${host.value}${p.path}" withMethod p.path
+      val url =  s"${host.value}${p.path}"
+
+      logger info s"requesting to $url with ${p.path}"
+
+      client url url withMethod p.path
```
Comment (iRhysBradbury) [inline client/src/main/scala/uk/gov/ons/addressIndex/client/AddressIndexClient.scala]: adding comment `//client.url(url).withMethod(p.path)`

**Comment by Mironor** [comment]
Query by uprn is broken (returns RichSearchResult instead of Json)

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchController.scala [hunk 14]
```diff
@@ -48,13 +43,15 @@ class SingleMatchController @Inject()(
...
     logger info("SingleMatch: Rendering Single Match Page")
-    val viewToRender = uk.gov.ons.addressIndex.demoui.views.html.singleMatch(
-      singleSearchForm = SingleMatchController.form,
-      warningMessage = None,
-      addressBySearchResponse = None,
-      classification = None)
     Future.successful(
-        Ok(viewToRender)
+      Ok(
+        uk.gov.ons.addressIndex.demoui.views.html.singleMatch(
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/controllers/SingleMatchController.scala]: There's no other references to it, no need for extra stack reference imo

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/partials/Address.scala.html [hunk 15]
```diff
@@ -0,0 +1,63 @@
+@*@import uk.gov.ons.addressIndex.model.server.response.AddressInformation*@
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/partials/Address.scala.html]: yes good spot

File: demo-ui/app/uk/gov/ons/addressIndex/demoui/views/partials/googleMap/googleMap.scala.html [hunk 16]
```diff
@@ -0,0 +1,27 @@
+@import uk.gov.ons.addressIndex.model.server.response.AddressInformation
```
Comment (iRhysBradbury) [inline demo-ui/app/uk/gov/ons/addressIndex/demoui/views/partials/googleMap/googleMap.scala.html]: Can be removed good spot

File: model/src/main/scala/uk/gov/ons/addressIndex/model/AddressIndexModel.scala [hunk 17]
```diff
@@ -3,13 +3,13 @@ package uk.gov.ons.addressIndex.model
 import java.util.UUID
 
 case class AddressIndexUPRNRequest(
-  format: AddressScheme,
+  format: Option[AddressScheme],
   uprn: BigInt,
   id: UUID
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/AddressIndexModel.scala]: It'll be used in logging info about requests - so we can track a request all the way through the system

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+  uprn: String,
+  lpi: JsValue,
+  paf: JsValue
+)
+
+object HybridIndex extends ElasticIndex[HybridIndex] {
+
+  val name: String = "hybrid"
+
+  implicit lazy val fmt = Json.format[HybridIndex]
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala]: Not sure it's used, whacked it in there just incase, not necessary IMO, can remove

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 19]
```diff
@@ -1,322 +1,196 @@
...
-      nag.townName, nag.postcodeLocator)
-  }
-
-  def fromNagAddress(other: NationalAddressGazetteerAddress): AddressResponseAddress = fromNagAddress(1.0f)(other)
-
+object AddressInformation {
+  implicit lazy val fmt: Format[AddressInformation] = Json.format[AddressInformation]
 }
 
+case class PAFWithFormat(formattedAddress: String, paf: PAF)
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: 22 fields :(

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 20]
```diff
@@ -1,322 +1,196 @@
...
+object Model {
 
-object AddressByUprnResponseContainer {
-  implicit val addressByUprnResponseContainerFormat: Format[AddressByUprnResponseContainer] =
-    Json.format[AddressByUprnResponseContainer]
-}
+  case class HybridResponse(
+    uprn: String,
+    lpi: Option[Seq[NationalAddressGazetteer]],
+    paf: Option[Seq[PostcodeAddressFile]],
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: The format param dictates the columns selected from elastic (we dont want to throw loads of data over the network)
This is why the fields are Optional

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 21]
```diff
@@ -1,322 +1,196 @@
...
-  /**
-    * Empty tokens (when needed before address tokenization)
-    */
-  val empty = AddressTokens(
-    uprn = "",
-    buildingNumber = "",
-    postcode = ""
-  )
+object Results {
+  implicit lazy val trFmt: Format[CrfTokenResult] = Json.format[CrfTokenResult]
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: Implicit values aren't referenced, `fmt` is generally the used val name

I don't might changing it though, doesn't change the functionality

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 22]
```diff
@@ -1,322 +1,196 @@
...
-  * @param underlyingMaxScore maximum score in the elastic result
-  */
-case class AddressResponseAddress(
+case class AddressInformation(
   uprn: String,
-  formattedAddress: String,
-  paf: Option[AddressResponsePaf],
-  nag: Option[AddressResponseNag],
-  geo: Option[AddressResponseGeo],
+  paf: Option[Seq[PAFWithFormat]],
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: We need Options because of the hybrid query

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 23]
```diff
@@ -1,322 +1,196 @@
...
+  nag: Option[Seq[NAGWithFormat]],
   underlyingScore: Float,
   underlyingMaxScore: Float
 )
 
-object AddressResponseAddress {
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: I moved it to a more OO model

`as an <X> I know how to be transformed to <Y>`

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 24]
```diff
@@ -380,161 +244,59 @@ case class AddressResponseNag(
...
-  * @param easting   easting
-  * @param northing  northing
-  */
-case class AddressResponseGeo(
+case class GEO(
   latitude: Double,
   longitude: Double,
-  easting: Int,
-  northing: Int
+  easting: Double,
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: They're `Doubles` in elastic, from what I can see (well `Double`s being stored as `String`s)

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 25]
```diff
@@ -380,161 +244,59 @@ case class AddressResponseNag(
...
   message: String
 )
 
-object AddressResponseStatus {
-  implicit lazy val addressResponseStatusFormat: Format[AddressResponseStatus] = Json.format[AddressResponseStatus]
+object Status {
+  implicit lazy val fmt: Format[Status] = Json.format[Status]
 }
 
-object OkAddressResponseStatus extends AddressResponseStatus(
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: Moved to Canned response helper, they're server canned responses :)

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala [hunk 26]
```diff
@@ -380,161 +244,59 @@ case class AddressResponseNag(
...
-  */
-case class AddressResponseError(
+case class Error(
   code: Int,
   message: String
 )
 
-object AddressResponseError {
-  implicit lazy val addressResponseErrorFormat: Format[AddressResponseError] = Json.format[AddressResponseError]
+object Error {
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: The context here is; `uk.gov.ons.addressIndex.model.server.response.AddressResponse` 
packages provide context generally - in that there's `Error`s
So the full context is: `uk.gov.ons.addressIndex.model.server.response.AddressResponse.Error`
Which seems sufficient; do you agree?

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala (hunk 18 above)
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala]: Ah, most of this file is legacy, managed to strip it right back to minimal stuff

File: model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala (hunk 26 above)
Comment (Mironor) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/server/response/AddressResponse.scala]: That would imply that you'd need to provide this context `AddressResponse.Error` every time that you are using this class. Thus this would apply artificial restrictions that nobody is forced to follow and will simply forget. By adding the prefix `AddressResponse` to the class, the context becomes mandatory.

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+    * This is needed to directly transform a collection of objects returned by Elastic
+    * request into a collection of PAF addresses
+    */
+  implicit object HybridResponseHitAs extends HitAs[HybridIndex] {
+    import Fields._
+
+    override def as(hit: RichSearchHit): HybridIndex = {
+      def map(key: String): String = hit.stringValue(key)
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala]: Suggestions?

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+  * @param addresses fetched addresses
+  * @param maxScore maximum score
+  */
+case class HybridResults(
+  addresses: Seq[HybridIndex],
+  maxScore: Float
+)
+
+object HybridResults {
+  implicit lazy val fmt = Json.format[HybridResults]
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala]: AddressController.scala

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala [hunk 29]
```diff
@@ -0,0 +1,100 @@
+package uk.gov.ons.addressIndex.model.db.index
+
+import uk.gov.ons.addressIndex.model.server.response._
+
+case class NationalAddressGazetteer(
```
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/NationalAddressGazetteer.scala]: Maybe we should move these into 
`uk.gov.addressIndex.model.db.index.format` to provide context?

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/PostcodeAddressFile.scala (hunk 10 above)
Comment (iRhysBradbury) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/PostcodeAddressFile.scala]: Maybe we should move these into
uk.gov.addressIndex.model.db.index.format to provide context?

File: server/app/uk/gov/ons/addressIndex/server/controllers/AddressController.scala [hunk 30]
```diff
@@ -1,106 +1,131 @@
...
     }
   }
 
   /**
-    * Address query API
-    *
-    * @param input  the address query
-    * @param format requested format of the query (paf/nag)
-    * @return Json response with addresses information
+    * @param input
```
Comment (iRhysBradbury) [inline server/app/uk/gov/ons/addressIndex/server/controllers/AddressController.scala]: Scaladoc no longer valid

File: model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala (hunk 27 above)
Comment (Mironor) [inline model/src/main/scala/uk/gov/ons/addressIndex/model/db/index/HybridIndex.scala]: You may just inline it, it will be fine

File: tagger/README.md (hunk 2 above)
Comment (iRhysBradbury) [inline tagger/README.md]: This is a readme.md file :)
